src/analysis.rs

```diff
@@ -27,358 +23,832 @@ pub fn generate_blackman(size: usize) -> impl Iterator<Item = f32> {
+        // imaginary component:
+        // cf. third_party/blink/renderer/modules/webaudio/realtime_analyser.h
+        // ```
+        // // Blow away the packed nyquist component.
+        // imag[0] = 0;
+        // ```
+        // In our case, it seems we can thus just ignore the Nyquist information
+        // and take the DC bin as it is
+
+        let normalize_factor = 1. / fft_size as f32;
+
+        last_fft_output
             .iter_mut()
             .zip(output.iter())
-            .for_each(|(p, c)| {
-                *p = smoothing_time_constant * *p + (1. - smoothing_time_constant) * c.norm()
+            .for_each(|(o, c)| {
+                let norm = c.norm() * normalize_factor;
+                let value = smoothing_time_constant * *o + (1. - smoothing_time_constant) * norm;
+                *o = if value.is_finite() { value } else { 0. };
+            });
+    }
+
+    pub fn get_float_frequency_data(&mut self, dst: &mut [f32], current_time: f64) {
+        let frequency_bin_count = self.frequency_bin_count();
+
+        // [spec] If another call to getByteFrequencyData() or getFloatFrequencyData()
+        // occurs within the same render quantum as a previous call, the current
+        // frequency data is not updated with the same data. Instead, the previously
+        // computed data is returned.
+        if current_time != self.last_fft_time {
+            self.compute_fft();
+            self.last_fft_time = current_time;
+        }
+
+        // [spec] Write the current frequency data into array. If array’s byte
+        // length is less than frequencyBinCount, the excess elements will be
+        // dropped. If array’s byte length is greater than the frequencyBinCount
+        let len = dst.len().min(frequency_bin_count);
+
+        // Convert to dB.
+        dst.iter_mut()
+            .take(len)
+            .zip(self.last_fft_output.iter())
+            .for_each(|(v, b)| *v = 20. * b.log10());
+    }
+
+    pub fn get_byte_frequency_data(&mut self, dst: &mut [u8], current_time: f64) {
+        let frequency_bin_count = self.frequency_bin_count();
+        let min_decibels = self.min_decibels() as f32;
+        let max_decibels = self.max_decibels() as f32;
+
+        // [spec] If another call to getByteFrequencyData() or getFloatFrequencyData()
+        // occurs within the same render quantum as a previous call, the current
+        // frequency data is not updated with the same data. Instead, the previously
+        // computed data is returned.
+        if current_time != self.last_fft_time {
+            self.compute_fft();
+            self.last_fft_time = current_time;
+        }
+
+        // [spec] Write the current frequency data into array. If array’s byte
+        // length is less than frequencyBinCount, the excess elements will be
+        // dropped. If array’s byte length is greater than the frequencyBinCount
+        let len = dst.len().min(frequency_bin_count);
```

Comment:
`zip` will stop when the end of either buffer is reached, so I don't think `take` is necessary